### PR TITLE
Testing image operations on snapshot mirrored cluster

### PR DESCRIPTION
**Pull request by manasagowri**
Automating test case - CEPH-83574857 - Configure two-way rbd-mirror (replicated and ec pool) on
    Stand alone CEPH cluster on image with snapshot based mirroring and perform
    image operations like expand/shrink/remove from primary and secondary clusters.

Test Case Flow
    1. Bootstrap two CEPH clusters and create a pool and 10 images each for replicated and ec pools
    2. Setup snapshot based mirroring in between these clusters.
    3. Run IOs on the Primary cluster for the image.
    4. Expand the image while IO is in progress in primary cluster
    5. Shrink the image while IO is in progress in primary cluster
    6. Remove the image while IO is in progress in primary cluster
    7. Try expand/shrink/remove operation from secondary cluster

Success Logs - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-SM709Q/

### CODE REVIEW

File: tests/rbd_mirror/test_image_operations_on_snap_mirroring.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    """
+    for pool_type in kw.get("pool_types"):
+        config = kw.get("config", {}).get(pool_type)
+        multi_pool_config = getdict(config)
+        for pool, pool_config in multi_pool_config.items():
+            multi_image_config = getdict(pool_config)
+            for image, image_config in multi_image_config.items():
+                image_size = image_config.get("size")
+
+                log.info(
```
Comment (sunilangadi2) [inline tests/rbd_mirror/test_image_operations_on_snap_mirroring.py:76]: we can iterate line 76 to 110 inside for loop for `primary` and `secondary` for code optimization.

**Review by sunilangadi2 — CHANGES_REQUESTED** [review]
I think some of it's related operation test case is there in existing modules there we should use large config with multi images,
also if it's tier-3 can we keep it in separate suite as we increase pool count and image count going with baremetal config?

File: tests/rbd_mirror/test_image_operations_on_snap_mirroring.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        log.error(
+            f"Image {image} shrink failed in {cluster_type} cluster with error {out} {err}"
+        )
+        return 1
+    elif (
+        cluster_type == "secondary"
+        and "new size is equal to original size" not in out + err
+    ):
+        log.error(
+            f"Image {image} shrink did not fail as expected in {cluster_type} cluster"
```
[the remark this answers is not in the document]
Comment (manasagowri) [reply tests/rbd_mirror/test_image_operations_on_snap_mirroring.py:42]: Done.

File: tests/rbd_mirror/test_image_operations_on_snap_mirroring.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    if cluster_type == "primary" and "100% complete" not in out + err:
+        log.error(
+            f"Image {image} remove failed in {cluster_type} cluster with error {out} {err}"
+        )
+        return 1
+    elif (
+        cluster_type == "secondary" and "mirrored image is not primary" not in out + err
+    ):
+        log.error(
+            f"Image {image} remove did not fail as expected in {cluster_type} cluster"
```
[the remark this answers is not in the document]
Comment (manasagowri) [reply tests/rbd_mirror/test_image_operations_on_snap_mirroring.py:57]: Done.

File: tests/rbd_mirror/test_image_operations_on_snap_mirroring.py (hunk 1 above)
Comment (manasagowri) [reply tests/rbd_mirror/test_image_operations_on_snap_mirroring.py:76]: Done.

**Comment by manasagowri** [comment]
@sunilangadi2 Addressed all review comments. Please review
Latest success logs - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-HW3R5P/

File: tests/rbd_mirror/test_image_operations_on_snap_mirroring.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            k for k in kw.get("ceph_cluster_dict").keys() if k != cluster_name
+        ][0]
+        pool_types = mirror_obj.get(cluster_name).get("pool_types")
+        rbd_obj = mirror_obj.get(cluster_name).get("rbd")
+        sec_obj = mirror_obj.get(sec_cluster_name).get("rbd")
+        ret_val = test_image_operations(
+            rbd_obj=rbd_obj, sec_obj=sec_obj, pool_types=pool_types, **kw
+        )
+        if ret_val:
+            log.error("Testing toggle journaling feature failed")
```
Comment (sunilangadi2) [inline tests/rbd_mirror/test_image_operations_on_snap_mirroring.py:151]: looks like error message is not related to this test case.

File: tests/rbd_mirror/test_image_operations_on_snap_mirroring.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    except Exception as e:
+        log.error(
+            f"Testing verify image operations from primary and secondary cluster\
```
Comment (sunilangadi2) [inline tests/rbd_mirror/test_image_operations_on_snap_mirroring.py:154]: keep any one either Testing or verifying, sentence confusing.

**Review by sunilangadi2 — CHANGES_REQUESTED** [review]

**Comment by sunilangadi2** [comment]
In the logs it says as below no where message that expand shrink remove is success or failed,
can you please make actions related info in the logs?
```2023-11-09 11:42:18,204 (cephci.test_image_operations_on_snap_mirroring) [INFO] - cephci.tests.rbd_mirror.test_image_operations_on_snap_mirroring.py:84 - Testing image expand/shrink/remove on primary cluster for image rep_image_UJrIV0
2023-11-09 11:42:18,206 (cephci.test_image_operations_on_snap_mirroring) [INFO] - cephci.ceph.ceph.py:1557 - Running command rbd resize rep_pool_XKZJR0/rep_image_UJrIV0  --size 2G on 10.0.210.103 timeout 600
2023-11-09 11:42:18,839 (cephci.test_image_operations_on_snap_mirroring) [INFO] - cephci.ceph.ceph.py:1557 - Running command rbd resize rep_pool_XKZJR0/rep_image_UJrIV0  --size 1G --allow-shrink on 10.0.210.103 timeout 600
2023-11-09 11:42:20,050 (cephci.test_image_operations_on_snap_mirroring) [INFO] - cephci.ceph.ceph.py:1557 - Running command rbd rm rep_pool_XKZJR0/rep_image_UJrIV0  on 10.0.210.103 timeout 600```

File: tests/rbd_mirror/test_image_operations_on_snap_mirroring.py (hunk 4 above)
Comment (manasagowri) [reply tests/rbd_mirror/test_image_operations_on_snap_mirroring.py:151]: done.

File: tests/rbd_mirror/test_image_operations_on_snap_mirroring.py (hunk 5 above)
Comment (manasagowri) [reply tests/rbd_mirror/test_image_operations_on_snap_mirroring.py:154]: done.

**Review by sunilangadi2 — APPROVED** [review]
changes LGTM.

**Pull request merged**